src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs

```diff
@@ -104,6 +125,195 @@ internal async Task<bool> ScheduleDurableTaskEvents(OrchestrationInvocationResul
+                    task = this.context.CallSubOrchestratorWithRetryAsync(action.FunctionName, action.RetryOptions, action.InstanceId, action.Input);
+                    break;
+                case AsyncActionType.CallEntity:
+                    {
+                        var entityId = EntityId.GetEntityIdFromSchedulerId(action.InstanceId);
+                        task = this.context.CallEntityAsync(entityId, action.EntityOperation, action.Input);
+                        break;
+                    }
+
+                case AsyncActionType.SignalEntity:
+                    {
+                        // We do not add a task because this is 'fire and forget'
+                        var entityId = EntityId.GetEntityIdFromSchedulerId(action.InstanceId);
+                        this.context.SignalEntity(entityId, action.EntityOperation, action.Input);
+                        break;
+                    }
+
+                case AsyncActionType.ScheduledSignalEntity:
+                    {
+                        // We do not add a task because this is 'fire and forget'
+                        var entityId = EntityId.GetEntityIdFromSchedulerId(action.InstanceId);
+                        this.context.SignalEntity(entityId, action.FireAt, action.EntityOperation, action.Input);
+                        break;
+                    }
+
+                case AsyncActionType.ContinueAsNew:
+                    this.context.ContinueAsNew(action.Input);
+                    break;
+                case AsyncActionType.WaitForExternalEvent:
+                    task = this.context.WaitForExternalEvent<object>(action.ExternalEventName);
+                    break;
+                case AsyncActionType.CallHttp:
+                    task = this.context.CallHttpAsync(action.HttpRequest);
+                    break;
+                default:
+                    break;
+            }
+
+            return task;
+        }
+
+        /// <summary>
+        /// Replays the orchestration execution from an OOProc SDK in .NET.
+        /// This is the V2 implementation of that replay procedure.
+        /// </summary>
+        /// <param name="actionSet">The OOProc actions payload.</param>
+        /// <returns>An awaitable Task that completes once replay completes.</returns>
+        private async Task ProcessAsyncActionsV2(AsyncAction[][] actionSet)
+        {
+            foreach (AsyncAction[] actions in actionSet)
+            {
+                Task newTask;
+                AsyncAction topLevelAction = actions[0];
+
+                if (actions.Length == 1)
+                {
+                    newTask = this.InvokeAPIFromAction(topLevelAction);
+                }
+                else
+                {
+                    // If there's more than one action, we know we're looking at a compound Task / a task of tasks
+                    // such as Task.WhenAll or Task.WhenAny. Recursively, we construct that compound task.
```

Comment:
Is this always true? Will this logic work if someone writes the following orchestrator logic?
```python
task1 = context.call_activity("foo")
task2 = context.call_activity("bar")
yield task1 # two activities were scheduled, but we're only waiting for one of them
yield task2
```

Reply:
Each `yield`ed statement creates its own action array. So the snippet of code you provided would look like:

```
[ 
   [ CallActivityAction.Foo ],
   [ CallActivityAction.Bar ]
]
```
and so each of them will be detected as a single element `actions` array :)

Reply:
Perfect, thanks! :)